pkg/tbtcpg/moving_funds.go

```diff
@@ -0,0 +1,628 @@
+	taskLogger log.StandardLogger,
+	sourceWalletPublicKeyHash [20]byte,
+	targetWalletsCommitmentHash [32]byte,
+) ([][20]byte, error) {
+	taskLogger.Infof(
+		"commitment already submitted; retrieving committed target wallets",
+	)
+
+	blockCounter, err := mft.chain.BlockCounter()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get block counter: [%w]",
+			err,
+		)
+	}
+
+	currentBlockNumber, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get current block number: [%w]",
+			err,
+		)
+	}
+
+	filterLookBackSeconds := uint64(MovingFundsCommitmentLookBackPeriod.Seconds())
+	filterLookBackBlocks := filterLookBackSeconds / uint64(mft.chain.AverageBlockTime().Seconds())
+	filterStartBlock := currentBlockNumber - filterLookBackBlocks
+
+	filter := &tbtc.MovingFundsCommitmentSubmittedEventFilter{
+		StartBlock:          filterStartBlock,
+		WalletPublicKeyHash: [][20]byte{sourceWalletPublicKeyHash},
+	}
+
+	events, err := mft.chain.PastMovingFundsCommitmentSubmittedEvents(filter)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past moving funds commitment submitted events: [%w]",
+			err,
+		)
+	}
+
+	// Moving funds commitment can be submitted only once for a given wallet.
+	// Check just in case.
+	if len(events) != 1 {
+		return nil, fmt.Errorf(
+			"unexpected number of moving funds commitment submitted events: [%v]",
+			len(events),
+		)
+	}
+
+	targetWallets := events[0].TargetWallets
+
+	// Make sure all the target wallets are Live.
+	for _, targetWallet := range targetWallets {
+		targetWalletChainData, err := mft.chain.GetWallet(targetWallet)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to get wallet data for target wallet [0x%x]: [%w]",
+				targetWallet,
+				err,
+			)
+		}
+
+		if targetWalletChainData.State != tbtc.StateLive {
+			return nil, fmt.Errorf(
+				"%w: [0x%x]",
+				ErrTargetWalletNotLive,
+				targetWallet,
+			)
+		}
+	}
+
+	// Just in case check if the hash of the target wallets matches the moving
+	// funds target wallets commitment hash.
+	packedWallets := []byte{}
+	for _, wallet := range targetWallets {
+		packedWallets = append(packedWallets, wallet[:]...)
+	}
+
+	calculatedHash := crypto.Keccak256(packedWallets)
+
+	if !bytes.Equal(calculatedHash, targetWalletsCommitmentHash[:]) {
+		return nil, ErrWrongCommitmentHash
+	}
+
+	return targetWallets, nil
+}
+
+// GetWalletMembersInfo returns the wallet member IDs based on the provided
+// wallet operator addresses. Additionally, it returns the position of the
+// moving funds task execution operator on the list.
+func (mft *MovingFundsTask) GetWalletMembersInfo(
+	walletOperators []chain.Address,
+	executingOperator chain.Address,
+) ([]uint32, uint32, error) {
+	walletMemberIDs := make([]uint32, 0)
+	walletMemberIndex := 0
+
+	for index, operatorAddress := range walletOperators {
+		operatorID, err := mft.chain.GetOperatorID(operatorAddress)
```

Comment:
This will always make `100` calls to the ETH node. Right now, we have 25 distinct operators. That means we can reduce the number of calls ~4x by introducing a simple local cache in this function.

Long-term, we can introduce a global cache at the `ProposalGenerator` level that will be reused across different tasks. We can add a TODO in order to capture that.

Reply:
Added a local cache (a map) in 99aeec51cec3591053f56f25dff079518b3ad1d8.